src/main/java/edu/wisc/my/keyvalue/controller/KeyValueStoreController.java

```diff
@@ -57,6 +66,102 @@ public void setUsernameAttr(String attr) {
+        String filterHeaderValue = request.getHeader(prefixAttr); 
+        authorized = filterHeaderValue != null;
+        prefix += ":" + filterHeaderValue;
+      }
+      
+      if(authorized) {
+        //security check success
+        
+        if(!isJSONValid(valueJson)) {
+          logger.error("Invalid request, json not valid");
+          response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+          return;
+        }
+        
+        logger.trace("Setting prefix: " + prefix + " key: " + key);
+        keyValueService.setValue(prefix, key, valueJson);
+        
+        // write response
+        try {
+          response.getWriter().write(valueJson);
+          response.setContentType("application/json");
+          response.setStatus(HttpServletResponse.SC_OK);
+        } catch (IOException e) {
+          logger.error("Issues happened while trying to write json", e);
+          response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+        }
+        
+        logger.info("user " + username + " wrote to scope " + scope + " and key " + key + " with value " + valueJson);
+      } else {
+        logger.error("User " + username + " attempted to PUT to " + scope + " but doesn't have access, shame!");
+        response.setStatus(HttpServletResponse.SC_FORBIDDEN);
+      }
+      
+    }
+    
+    @RequestMapping(value="/{scope}/{key}", method=RequestMethod.GET)
+    public @ResponseBody void getScopedKeyValue(HttpServletRequest request, HttpServletResponse response, @PathVariable String scope, @PathVariable String key) throws IOException {
+      boolean scopeUserBased = Boolean.parseBoolean(env.getRequiredProperty("scope." + scope + ".byUser"));
+      
+      if(scopeUserBased) {
+        String property = env.getProperty("scope." + scope + ".prefixAttribute");
+        String propHeaderValue = request.getHeader(property);
+        if(propHeaderValue != null) {
+          scope += ":" + propHeaderValue;
+        } else {
+          //there was a property for this scope, but the proper header was not set
+          logger.error(ACCESS_ERROR);
+          response.setStatus(HttpServletResponse.SC_FORBIDDEN);
+          return;
+        }
+      } else {
+        logger.trace("global hit");
+      }
+      logger.trace("Searching for prefix " + scope + " key: " + key);
+      String value = keyValueService.getValue(scope, key);
+      try {
+          if(isJSONValid(value)) {
+            logger.trace("Got something for scope : " + scope + ", key : " + key + ", value : " + value);
+            //valid json, cool, write it
+            response.getWriter().write(value);
+            response.setContentType("application/json");
+            response.setStatus(HttpServletResponse.SC_OK);
+          }
+          else {
+            logger.trace("Got nothing for scope : " + scope + ", key : " + key);
+            response.setStatus(HttpServletResponse.SC_OK);
```

Comment:
204 can be wonky, particularly in IE. It's the correct code per the specification, but it's not consistently understood.
There is much debate over what is the appropriate status code for a null response to a REST resource. I've used 404 often.